File: jbake-core/src/main/java/org/jbake/parser/ParserContext.java
package org.jbake.parser;

import org.jbake.app.configuration.JBakeConfiguration;
import org.jbake.model.DocumentModel;

import java.io.File;
import java.time.Instant;
import java.util.List;

public class ParserContext {
    private final File file;
    private final List<String> fileLines;
    private final JBakeConfiguration config;
    private final boolean hasHeader;
    private final DocumentModel documentModel;

    public ParserContext(
            File file,
            List<String> fileLines,
            JBakeConfiguration config,
            boolean hasHeader) {
        this.file = file;
        this.fileLines = fileLines;
        this.config = config;
        this.hasHeader = hasHeader;
        this.documentModel = DocumentModel.createDefaultDocumentModel();
    }

    public File getFile() {
        return file;
    }

    public List<String> getFileLines() {
        return fileLines;
    }

    public JBakeConfiguration getConfig() {
        return config;
    }

    public DocumentModel getDocumentModel() {
        return documentModel;
    }

    public boolean hasHeader() {
        return hasHeader;
    }

    // short methods for common use
    public String getBody() {
        return documentModel.getBody();
    }

    public void setBody(String str) {
        documentModel.setBody(str);
    }

    public Instant getDate() {
        return getDocumentModel().getDate();
    }

    public void setDate(Instant date) {
        getDocumentModel().setDate(date);
    }

    public String getStatus() {
        if (getDocumentModel().getStatus() != null) {
            return getDocumentModel().getStatus();
        }
        return "";
    }

    public void setDefaultStatus() {
        getDocumentModel().setStatus(getConfig().getDefaultStatus());
    }

    public String getType() {
        if (getDocumentModel().getType() != null) {
            return getDocumentModel().getType();
        }
        return "";
    }

    public void setDefaultType() {
        getDocumentModel().setType(getConfig().getDefaultType());
    }

    public Object getTags() {
        return getDocumentModel().getTags();
    }

    public void setTags(String[] tags) {
        getDocumentModel().setTags(tags);
    }
}

File: jbake-core/src/main/java/org/jbake/parser/ParserContext.kt
package org.jbake.parser

import org.jbake.app.configuration.JBakeConfiguration
import org.jbake.model.DocumentModel
import java.io.File
import java.util.*

class ParserContext(
    val file: File,
    val fileLines: MutableList<String>,
    val config: JBakeConfiguration,
    private val hasHeader: Boolean
) {
    val documentModel: DocumentModel = DocumentModel.createDefaultDocumentModel()

    fun hasHeader(): Boolean = hasHeader

    var body: String
        get() = documentModel.body
        set(str) {
            documentModel.body = str
        }

    var date: Date?
        get() = this.documentModel.date
        set(date) { this.documentModel.date = date}

    val status: String
        get() = this.documentModel.status ?: ""

    fun setDefaultStatus() {
        this.documentModel.status = this.config.defaultStatus ?: ""
    }

    val type: String
        get() = this.documentModel.type

    fun setDefaultType() {
        this.documentModel.type = this.config.defaultType ?: ""
    }

    fun setDefaultAuthor() {
        if (this.documentModel.containsKey("author")) return

        val defaultAuthor = this.config.defaultAuthor
        if (!defaultAuthor.isNullOrEmpty())
            this.documentModel["author"] = defaultAuthor
    }

    // TODO: Possibly nullable
    val tags: List<String>
        get() = this.documentModel.tags

    fun setTags(tags: List<String>) {
        this.documentModel.tags = tags
    }
}
